File: WQChartsDemo/Objc/ViewController/RadialChartVC.m
// https://github.com/CoderWQYao/WQCharts-iOS
//
// RadialChartVC.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "RadialChartVC.h"

@interface RadialChartVC ()

@property (nonatomic, readonly) WQRadialChartView* radialChartView;

@end

@implementation RadialChartVC

- (WQRadialChartView *)radialChartView {
    return (WQRadialChartView*)self.chartViewRef;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    self.radialChartView.frame = self.chartContainer.bounds;
}

- (void)chartViewDidCreate:(WQRadialChartView *)chartView {
    chartView.drawDelegate = self;
    chartView.rotationDelegate = self;
}

- (void)configChartOptions {
    [super configChartOptions];
    __weak typeof(self) weakSelf = self;
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"Padding")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadialChartView* chartView = weakSelf.radialChartView;
        CGFloat padding = selection != 0 ? 30 : 0;
        chartView.padding = UIEdgeInsetsMake(padding, padding, padding, padding);
    })];
    
    [self.optionsView addItem:ListCell.new
     .setTitle(@"Angle")
     .addItem(SliderCell.new
              .setSliderValue(0,360,360)
              .setOnValueChange(^(SliderCell* cell,float value) {
        WQRadialChartView* chartView = weakSelf.radialChartView;
        WQRadialChart* chart = chartView.chartAsRadial;
        chart.angle = value;
        [chartView redraw];
    }))];
    
    [self.optionsView addItem:ListCell.new
     .setTitle(@"Rotation")
     .addItem(SliderCell.new
              .setSliderValue(0,360,0)
              .setOnValueChange(^(SliderCell* cell,float value) {
        WQRadialChartView* chartView = weakSelf.radialChartView;
        WQRadialChart* chart = chartView.chartAsRadial;
        chart.rotation = value;
        [chartView redraw];
    }))];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"CounterClockwise")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(0)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQRadialChartView* chartView = weakSelf.radialChartView;
        WQRadialChart* chart = (WQRadialChart*)[chartView valueForKey:@"chart"];
        chart.direction = selection != 0 ? WQPieChartDirectionCounterClockwise : WQPieChartDirectionClockwise;
        [chartView redraw];
    })];
    
}

#pragma mark - ChartViewDrawDelegate

- (void)chartViewWillDraw:(WQChartView *)chartView inRect:(CGRect)rect context:(CGContextRef)context {
    
}

- (void)chartViewDidDraw:(WQChartView *)chartView inRect:(CGRect)rect context:(CGContextRef)context {
    
}

#pragma mark - RadialChartViewRotationDelegate

- (void)radialChartView:(WQRadialChartView *)radialChartView rotationDidChange:(CGFloat)rotation translation:(CGFloat)translation {
    [self updateSliderValue:rotation forKey:@"Rotation" atIndex:0];
    NSLog(@"radialChartView: rotationDidChange: %f translation: %f", rotation, translation);
}

#pragma mark - Animation

- (void)appendAnimationKeys:(NSMutableArray<NSString *> *)animationKeys {
    [super appendAnimationKeys:animationKeys];
    [animationKeys addObject:@"Padding"];
    [animationKeys addObject:@"Clip"];
    [animationKeys addObject:@"Angle"];
    [animationKeys addObject:@"Rotation"];
}

- (void)prepareAnimationOfChartViewForKeys:(NSArray<NSString*>*)keys {
    [super prepareAnimationOfChartViewForKeys:keys];
    
    WQRadialChartView* chartView = self.radialChartView;
    
    if ([keys containsObject:@"Padding"]) {
        RadioCell* paddingCell = [self findRadioCellForKey:@"Padding"];
        CGFloat paddingValue = paddingCell.selection == 0 ? 30 : 0;
        UIEdgeInsets padding = UIEdgeInsetsMake(paddingValue, paddingValue, paddingValue, paddingValue);
        chartView.transformPadding = [[WQTransformUIEdgeInsets alloc] initWithFrom:chartView.padding to:padding];
        paddingCell.selection = paddingCell.selection == 0 ? 1 : 0;
    }
    
    if ([keys containsObject:@"Clip"] && chartView.graphicAsRadial) {
        WQRadialGraphic* graphic = chartView.graphicAsRadial;
        CGPoint center = graphic.center;
        CGFloat radius = graphic.radius;
        chartView.transformClipRect = [[WQTransformCGRect alloc] initWithFrom:(CGRect){center,CGSizeZero} to:CGRectMake(center.x - radius, center.y - radius, radius * 2, radius * 2)];
    }
    
    WQRadialChart* chart = chartView.chartAsRadial;
    
    if ([keys containsObject:@"Angle"]) {
        CGFloat angle = [NSNumber randomCGFloatFrom:0 to:360 + 90];
        angle = MIN(angle, 360);
        chart.transformAngle = [[WQTransformCGFloat alloc] initWithFrom:chart.angle to:angle];
    }
    
    if ([keys containsObject:@"Rotation"]) {
        chart.transformRotation = [[WQTransformCGFloat alloc] initWithFrom:chart.rotation to:[NSNumber randomCGFloatFrom:0 to:360]];
    }
}

#pragma mark - AnimationDelegate

- (void)animation:(WQAnimation *)animation progressDidChange:(CGFloat)progress {
    [super animation:animation progressDidChange:progress];
    
    if (animation.transformable == self.radialChartView) {
        WQRadialChart* chart = self.radialChartView.chartAsRadial;
        [self updateSliderValue:chart.angle forKey:@"Angle" atIndex:0];
        [self updateSliderValue:chart.rotation forKey:@"Rotation" atIndex:0];
    }
    
}

@end

File: WQChartsDemo/Swift/ViewController/RadialChartVC.swift
// https://github.com/CoderWQYao/WQCharts-iOS
//
// RadialChartVC.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class RadialChartVC<T: RadialChartView>: BaseChartVC<T>, ChartViewDrawDelegate, RadialChartViewRotationDelegate {
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        chartView.frame = chartContainer.bounds
    }
    
    override func chartViewDidCreate(_ chartView: T) {
        super.chartViewDidCreate(chartView)
        chartView.drawDelegate = self
        chartView.rotationDelegate = self
    }
    
    override func configChartOptions() {
        super.configChartOptions()
        
        optionsView.addItem(RadioCell()
            .setTitle("Padding")
            .setOptions(["OFF","ON"])
            .setSelection(1)
            .setOnSelectionChange({[weak chartView](cell, selection) in
                guard let chartView = chartView else {
                    return
                }
                let paddingValue: CGFloat = selection != 0 ? 30 : 0
                chartView.padding = UIEdgeInsets(top: paddingValue, left: paddingValue, bottom: paddingValue, right: paddingValue)
            })
        )
        
        optionsView.addItem(ListCell()
            .setTitle("Angle")
            .addItem(SliderCell()
                .setSliderValue(0,360,360)
                .setOnValueChange({[weak chartView](cell, value) in
                    guard let chartView = chartView else {
                        return
                    }
                    let chart = chartView.chartAsRadial
                    chart.angle = value
                    chartView.redraw()
                })
            )
        )
        
        optionsView.addItem(ListCell()
            .setTitle("Rotation")
            .addItem(SliderCell()
                .setSliderValue(0,360,0)
                .setOnValueChange({[weak chartView](cell, value) in
                    guard let chartView = chartView else {
                        return
                    }
                    let chart = chartView.chartAsRadial
                    chart.rotation = value
                    chartView.redraw()
                })
            )
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("CounterClockwise")
            .setOptions(["OFF","ON"])
            .setSelection(0)
            .setOnSelectionChange({[weak chartView](cell, selection) in
                guard let chartView = chartView else {
                    return
                }
                let chart = chartView.chartAsRadial
                chart.direction = selection != 0 ? .CounterClockwise : .Clockwise
                chartView.redraw()
            })
        )
        
    }
    
    // MARK: - ChartViewDrawDelegate
    
    func chartViewWillDraw(_ chartView: ChartView, inRect rect: CGRect, context: CGContext) {
        
    }
    
    func chartViewDidDraw(_ chartView: ChartView, inRect rect: CGRect, context: CGContext) {
        
    }
    
    
    // MARK: - RadialChartViewRotationDelegate
    
    func radialChartView(_ radialChartView: RadialChartView, rotationDidChange rotation: CGFloat, translation: CGFloat) {
        updateSliderValue(rotation, forKey: "Rotation", atIndex: 0)
        print("radialChartView:", "rotationDidChange:", rotation, "translation:", translation)
    }
    
    
    // MARK: - Animation
    
    override func appendAnimationKeys(_ animationKeys: NSMutableArray) {
        super.appendAnimationKeys(animationKeys)
        animationKeys.add("Padding")
        animationKeys.add("Clip")
        animationKeys.add("Angle")
        animationKeys.add("Rotation")
    }
    
    override func prepareAnimationOfChartView(forKeys keys: [String]) {
        super.prepareAnimationOfChartView(forKeys: keys)
        
        let chartView = self.chartView
        
        if keys.contains("Padding") {
            let paddingCell = findRadioCellForKey("Padding")!
            let paddingValue: CGFloat
            if paddingCell.selection == 0  {
                paddingValue = 30
            } else {
                paddingValue = 0
            }
            let padding = UIEdgeInsets(top: paddingValue, left: paddingValue, bottom: paddingValue, right: paddingValue)
            chartView.paddingTween = ChartUIEdgeInsetsTween(chartView.padding, padding)
            paddingCell.selection = paddingCell.selection == 0 ? 1 : 0
        }
        
        if keys.contains("Clip"), let graphic = chartView.graphicAsRadial {
            let center = graphic.center
            let radius = graphic.radius
            chartView.clipRectTween = ChartCGRectTween(CGRect(origin: center, size: .zero), CGRect(x: center.x - radius, y: center.y - radius, width: radius * 2, height: radius * 2))
        }
        
        let chart = chartView.chartAsRadial
        
        if keys.contains("Angle") {
            var angle = CGFloat.random(in: 0...360 + 90)
            angle = min(angle, 360)
            chart.angleTween = ChartCGFloatTween(chart.angle, angle)
        }
        
        if keys.contains("Rotation") {
            chart.rotationTween = ChartCGFloatTween(chart.rotation, CGFloat.random(in: 0...360))
        }
    }
    
    // MARK: - AnimationDelegate
    
    override func animation(_ animation: ChartAnimation, progressDidChange progress: CGFloat) {
        super.animation(animation, progressDidChange: progress)
        
        if chartView.isEqual(animation.animatable) {
            let chart = chartView.chartAsRadial
            updateSliderValue(chart.angle, forKey: "Angle", atIndex: 0)
            updateSliderValue(chart.rotation, forKey: "Rotation", atIndex: 0)
        }
    }
    
}
